rfcs/atomic_record_fields.md

```diff
@@ -0,0 +1,240 @@
+```
+
+Users would want to have fields where some mutable fields are
+atomic. This would allow a more efficient memory representation than
+the current one, which can make a qualitative difference for low-level
+concurrent code implementing important concurrent data structures.
+
+
+## The easy part
+
+From a distance, this sounds easy: in addition to `mutable`, just add
+an `atomic` keyword in field declarations.
+
+```ocaml
+type t = {
+  id : int;
+  atomic mutable state : st;
+}
+```
+
+Then have `foo.state` perform an atomic read (strong /
+sequentially consistent), and `foo.state <- v` perform an atomic
+write.
+
+People might not be in favor of adding a new keyword, but we could
+handle this with a builtin attribute:
+
+```ocaml
+type t = {
+  id : int;
+  mutable state : st [@atomic];
+}
+```
+
+
+## The hard part
+
+The difficulty is how to expose to the users the other atomic
+operations : how to support atomic `exchange`, `compare_and_set`,
+`fetch_and_add`, etc?
+
+One approach would be to add new langage constructs for all of these:
+`foo.state <-> new_state`, `cas foo.state old_state new_state`,
+`foo.state += 2`, etc.
+
+But this will not do:
+
+- language designers are reluctant to add new keywords or builtin
+  constructs; `atomic` will require a bit of convincing, `cas` is
+  a no-go
+  
+- there is no limit to the amount of builtin constructs we would need
+  to add in this situation, if we want to add new atomic operations
+  (for example have `add_and_fetch` as well)
+  
+Even if we accept to make "atomic fields" a builtin construct, we want
+to expose atomic operations on them as *library functions* /
+primitives, not new language constructs.
+
+Another idea would be to add a magic way to see the field `state` as
+an `Atomic.t` value. But this is deeply incompatible with the OCaml
+memory representation, which has no support for internal pointers:
+except for the very first record field there is no way to expose
+a pointer to a specific field as a valid atomic value, except by
+copying the value into a new atomic, which has a completely different
+(and useless and inefficient) semantics.
+
+## The proposal: phantom-typed first-class atomic fields
+
+In addition to the `atomic` keyword (or builtin attribute) on record
+fields, we propose to add:
+
+1. a new type `('r, 'a) Atomic.Field.t`, which denotes a field of type
+  `'a` within a record of type `'r`. Its representation is just an
+  integer, the offset of the field within the record structure.
+  
+2. Atomic operations taking both a `'r` and a `('r, 'a) t` in this new
+   Atomic.Field module, which mimick/duplicate the operations on
+   `Atomic.t`.
+  
+3. A single builtin-in construct to construct `('r, 'a) Atomic.Field.t`
+   values out of atomic record field names: `[%atomic.field foo]`
+
+For example, if we have `type 'a t = { atomic mutable state : 'a st }`,
+then `[%atomic.field state]` is a polymorphic constant of type
+`('a t, 'a st) Atomic.Field.t`.
```

Comment:
I wonder how hard it would be to have a third type parameter, which would be instantiated to a different (opaque) type for every field of a record. This way, when two fields have the same type, it would be impossible to mistakenly access the wrong one.